docker/lib/dependabot/docker/version.rb

```diff
@@ -4,9 +4,33 @@
 
 module Dependabot
   module Docker
+    #
+    # A Docker version is generally a RubyGems compatible version number. This
+    # RubyGems compatible release is kept in `@release_part`.
+    #
+    # In the special case of Java, the release part has a variable number of
+    # components, and may be followed by an extra component called "update
+    # release". This is kept separately in `@update_part`. See
+    # https://www.oracle.com/java/technologies/javase/versioning-naming.html for
+    # a description of Java versions.
+    #
     class Version < Gem::Version
       def initialize(version)
-        super(version.tr("_", "."))
+        release_part, update_part = version.split("_", 2)
+
+        @release_part = Gem::Version.new(release_part)
+
+        @update_part = Gem::Version.new(update_part&.start_with?(/[0-9]/) ? update_part : 0)
```

Comment:
The current implementation would use RubyGems criteria which is indeed lexicographical order. That in practice works reasonably well (alfa < beta < pre < rc).

In any case, I'm not sure whether in the case of Docker we actually receive identifiers here, maybe there are already filtered out. There are already a few complicated regexes in the update checker to detect version tags, so I'd rather go with a minimal patch that fixes our current issue.

Reply:
I just checked and I think my recollection is correct: the update checker only compares numeric versions. All instantiations of `Version` in `UpdateChecker` go through this method first:

https://github.com/dependabot/dependabot-core/blob/5ae1a8c240a1cdc45545672200ac8f0f80323094/docker/lib/dependabot/docker/update_checker.rb#L315-L319

The `NAME_WITH_VERSION` regex is quite complicated but I checked empirically that `<version>` seems to only captures only the numeric parts:

https://github.com/dependabot/dependabot-core/blob/5ae1a8c240a1cdc45545672200ac8f0f80323094/docker/lib/dependabot/docker/update_checker.rb#L45-L58

So in the example you gave, `11.0.16_8-ea` and `11.0.16.8-jdk` would result in `Version.new("11.0.16_8")` and (another) `Version.new("11.0.16_8")` being instantiated from sorting/comparison, and they would be considered as equal (so no updates proposed).

I think the current code could be improved and the above logic probably belongs in the version class rather than in the `UpdateChecker`. But a custom version class for `Docker` was only extracted at https://github.com/dependabot/dependabot-core/pull/5734/files, so that kind of refactoring has not yet happened.

Reply:
Created a follow-up ticket at #5911!